### PR TITLE
Hw5 courier

### PR DESCRIPTION
**Pull request by TomaszNakonieczny80**

### CODE REVIEW

File: Courier.BusinessLayer/CarsSerice.cs [hunk 1]
```diff
@@ -6,17 +6,26 @@
 
 namespace Courier.BusinessLayer
 {
-    public class CarsService
+    public interface ICarsService
     {
+        void Add(Car car);
+    }
+    public class CarsService : ICarsService
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/CarsSerice.cs:13]: Add empty line between interface definition and class implementation

File: Courier.BusinessLayer/ParcelsService.cs [hunk 2]
```diff
@@ -6,15 +6,46 @@
 using Courier.BusinessLayer.Models;
 using Courier.DataLayer;
 using Courier.DataLayer.Models;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal;
 
 namespace Courier.BusinessLayer
 {
-    public class ParcelsService
+    public interface IParcelsService
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:14]: Very large interface. Consider splitting (and as it is generated based on ParcelService class, consider splitting this class as well).

File: Courier.BusinessLayer/ParcelsService.cs [hunk 3]
```diff
@@ -77,42 +108,94 @@ public CarParcel GetParcelId(int parcelId)
...
-            using (var context = new ParcelsDbContext())
+            using (var context = _dbContextFactoryMethod())
             {
-                var parcelsToBePosted = context.Parcels.Where(parcel => parcel.ParcelStatus == 0).ToList();
+                var parcelsToBePosted = context.Parcels.AsQueryable().Where(parcel => parcel.ParcelStatus == 0).ToList();
                 return parcelsToBePosted;
             }
         }
 
+        public List<Parcel> GetPostedParcels()
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:118]: This method and GetParcelsWaitingToBePosted can be made one method requiring with ParcelStatus parameter

File: Courier.BusinessLayer/ParcelsService.cs [hunk 4]
```diff
@@ -77,42 +108,94 @@ public CarParcel GetParcelId(int parcelId)
...
+                    .Include(user => user.Sender)
+                    .ThenInclude(address => address.Address)
+                    .Include(user => user.Recipient)
+                    .ThenInclude(address => address.Address)
+                    .Where(parcel => parcel.ParcelStatus == ParcelStatus.OnTheWay).ToList();
+                return parcels;
+            }
+        }
+
+        public void SetParcelsAsOnTheWay()
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:142]: Again - this method can be easily made more generic: 
```
public void UpdatePackagesStatus(PackageStatus currentStatus, PackageStatus newStatus)
{
    ...
}
```

File: Courier.BusinessLayer/ParcelsService.cs [hunk 5]
```diff
@@ -77,42 +108,94 @@ public CarParcel GetParcelId(int parcelId)
...
             foreach (var parcel in GetParcelsWaitingToBePosted())
             {
-                using (var context = new ParcelsDbContext())
+                using (var context = _dbContextFactoryMethod())
                 {
                     foreach (var carsAvailable in GetAvailableCars())
                     {
-                        var carAvailable = context.Cars.Where(car => car.Id == carsAvailable.Id).FirstOrDefault();
+                        var carAvailable = context.Cars.AsQueryable().Where(car => car.Id == carsAvailable.Id).FirstOrDefault();
+                        var distanceToParcel = Math.Round((Math.Sqrt((Math.Pow((parcel.ParcelLatitude - carAvailable.Latitude), 2)) + ((Math.Pow((parcel.ParcelLongitude - carAvailable.Longitude), 2)))) * 73));
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:178]: Move distance calculation to separate method.

File: Courier.BusinessLayer/ParcelsService.cs [hunk 6]
```diff
@@ -126,60 +209,84 @@ public List<Shipment> AttachDriverToParcel()
...
+                        
+                        if (context.CarParcels.AsQueryable().Where(parcel => parcel.Full == false && parcel.AvailableTime - parcel.TotalTravelTime >= 0)
+                            .OrderBy(parcel => parcel.DistanceToParcel)
+                            .FirstOrDefault(parcel => parcel.ParcelId == selectedParcel.Id) == null)
+                        {
+                            carId = null;
+                        }
+                        else
+                        {
+                            carId = context.CarParcels.AsQueryable().Where(parcel => parcel.Full == false && parcel.AvailableTime - parcel.TotalTravelTime >= 0)
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:233]: here you do exactly the same query as in line 225. It can be called only once with ease:

```
var carId = context.CarParcels
    .AsQueryable()
    .Where(parcel => parcel.Full == false && parcel.AvailableTime - parcel.TotalTravelTime >= 0)
    .OrderBy(parcel => parcel.DistanceToParcel)
    .FirstOrDefault(parcel => parcel.ParcelId == selectedParcel.Id)
    ?.CarId;
```

File: Courier.BusinessLayer/ParcelsService.cs [hunk 7]
```diff
@@ -126,60 +209,84 @@ public List<Shipment> AttachDriverToParcel()
...
-                    Update(parcel);
+                    if (carId != null)
+                    {
 
-                    selectedParcel.ParcelStatus = ParcelStatus.Posted;
-                    Update(selectedParcel);
+                        int parcelWeight = (int) selectedParcel.ParcelSize;
+                        //licze ile czasu dostepnego zostanie kurierowi po przypisaniu paczki
+                        double deltaTime = GetCarParcel(selectedParcel.Id).AvailableTime -
+                                           GetCarParcel(selectedParcel.Id).TotalTravelTime;
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:247]: Why you call `GetCarParcel()` method twice with the same parameters in a single line?

Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:247]: At the same time I dont get it - to calculate how much time there is still for the driver you substract two different fields from a single entity. But entire time left is connected with a collection of all the parcels, when time delivery of a single package is connected with a particular package.

In addition, the requirement was to plan a route as follows:
1. Driver leaves his starting location.
2. Driver collects all the packages on the shipping list,
3. Driver deliveres all the packages on the shipping list,
4. Drive heads back to starting location.

Here it looks more like Drivers grabs a package, delivers and then grabs next one, delivers and so on and so on...

File: Courier.BusinessLayer/ParcelsService.cs [hunk 8]
```diff
@@ -126,60 +209,84 @@ public List<Shipment> AttachDriverToParcel()
...
+                            
+
+                            carParcel.AvailableCapcity -= (uint) parcelWeight;
+                            carParcel.AvailableTime = deltaTime;
 
                             Update(carParcel);
                         }
+                        //oznaczam paczke jako wysłana/przypisana do samochodu
+
+                        var parcel = GetParcelId(selectedParcel.Id);
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:262]: Method tells GetParcelId() when it actually returns parcel based on it's Id. Thats misleading.
Additionally, you call GetParcelId(selectedParcel.Id) very very many times...

File: Courier.BusinessLayer/ParcelsService.cs [hunk 9]
```diff
@@ -126,60 +209,84 @@ public List<Shipment> AttachDriverToParcel()
...
-                            .OrderBy(parcel => parcel.Distance)
-                            .FirstOrDefault(parcel => parcel.ParcelId == selectedParcel.Id)
-                            .CarId;
-                    }
+                    //jezeli kurier jest juz zapakowany na full lub total czas odbioru i przesyłki przekaracza godziny pracy kuriera biore nastepnego najblizszego kuriera
 
-                    int parcelWeight = (int) selectedParcel.ParcelSize;
-                    //aktualizuje dostepna ładownosc samochodu po przypisaniu paczki, 
-                    foreach (var carParcelId in GetListCarParcelId(carId))
+                    using (var context = _dbContextFactoryMethod())
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:222]: Quite a lot happens inside this using. Do you think some of the code can be done with closed database connection?

File: Courier.BusinessLayer/Models/Shipment.cs [hunk 10]
```diff
@@ -7,7 +7,7 @@ namespace Courier.BusinessLayer.Models
     public class Shipment
     {
         public Guid ParcelNumber { get; set; }
-        public int DriverId { get; set; }
+        public int? DriverId { get; set; }
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/Models/Shipment.cs:10]: Why? Do you consider having shipments in the system without having driver assigned?

File: Courier.BusinessLayer/ParcelsService.cs [hunk 11]
```diff
@@ -189,18 +296,18 @@ public List<Shipment> AttachDriverToParcel()
             return shipementList;
         }
 
-        public List<int> GetListCarParcelId(int carId)
+        public List<int> GetListCarParcelId(int? carId)
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/ParcelsService.cs:299]: I wonder - why you change it to nullable? And as you did - whu dont you check if it's null? Unless you assume there can be a CarParcel without car being assigned, then fine. But it's smelly....

File: Courier.BusinessLayer/TimeService.cs [hunk 12]
```diff
@@ -4,15 +4,19 @@
...
+    public class TimeService : ITimeService
     {
         public DateTime currentTime()
         {
             var start = new DateTime(2020, 11, 1);
             var now = DateTime.Now;
 
             var different = now - start;
-            var timeSpan =  different.TotalSeconds * 60;
+            var timeSpan =  different.TotalSeconds * 60 *60;
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/TimeService.cs:19]: dev change possibly

File: Courier.BusinessLayer/UsersService.cs [hunk 13]
```diff
@@ -33,7 +52,7 @@ public bool CheckIfUserExist(string userEmail)
 
         public bool CheckIfPasswordIsCorrect(string userPassword)
         {
-            using (var context = new ParcelsDbContext())
+            using (var context = _dbContextFactoryMethod())
             {
                 var result = context.Users.FirstOrDefault(user => user.Password == userPassword);
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/UsersService.cs:57]: So, correct password it when any of the users in the entire system has one? Or maybe it should be a valid combination of user and password?

File: Courier.BusinessLayer/UsersService.cs [hunk 14]
```diff
@@ -46,7 +65,7 @@ public bool CheckIfPasswordIsCorrect(string userPassword)
 
         public bool CheckIfDriver(int userId)
```
Comment (Bulczi-CM) [inline Courier.BusinessLayer/UsersService.cs:66]: `public bool IsDriver(int userId)`

File: Courier.DataLayer/Models/CarParcel.cs [hunk 15]
```diff
@@ -8,12 +8,24 @@ public class CarParcel
     {
         public int Id { get; set; }
         public int ParcelId { get; set; }
-        public double ParcelLatitude { get; set; }
-        public double ParcelLongitude { get; set; }
+        public int ParcelSize { get; set;}
+        //public double ParcelLatitude { get; set; }
```
Comment (Bulczi-CM) [inline Courier.DataLayer/Models/CarParcel.cs:12]: uncomment if needed, throw away if not

File: CourierService/MenuItem.cs [hunk 16]
```diff
@@ -4,7 +4,11 @@
 
 namespace Courier
 {
-    class MenuItem
+    public interface IMenuItem
```
Comment (Bulczi-CM) [inline CourierService/MenuItem.cs:7]: Well, seems line this interface is empty, so it is not needed. MenuItem is kind of a model class - it carries data that Menu uses. So it should not be injected, and it dont need interface in that matter (unless in our case right now).

File: CourierService/Program.cs [hunk 17]
```diff
@@ -82,19 +110,52 @@ private void SetTimer()
...
-            var deltaMillsek = delta.TotalMilliseconds;
-            
-            Timer aTimer = new Timer();
-            aTimer.Elapsed += new ElapsedEventHandler(OnTimedEvent);
-         
-            aTimer.Interval = deltaMillsek/60;
-            aTimer.Enabled = true;
+            var deltaMilisec = delta.TotalMilliseconds;
+            var eightHoursMilisec = 28800000;
+            var eighteenHoursMilisec = 64800000;
```
Comment (Bulczi-CM) [inline CourierService/Program.cs:115]: Use const int for that.

Hint: to make it more readable use multiplication, then reviewer can actually check if there are no errors:
`const int eighteenHoursMilisec = 18 * 60 * 60 * 1000;`
Consts are calculated in compile time ;)

File: CourierService/Program.cs [hunk 18]
```diff
@@ -82,19 +110,52 @@ private void SetTimer()
...
+            Timer aTimerForStartDelivery = new Timer();
+            aTimerForStartDelivery.Elapsed += new ElapsedEventHandler(SetStatusAsOnTheWay);
+
+            aTimerForStartDelivery.Interval = (deltaMilisec + eightHoursMilisec) / 600;
+            aTimerForStartDelivery.Enabled = true;
+
+            Timer aTimerForStopDelivery = new Timer();
+            aTimerForStopDelivery.Elapsed += new ElapsedEventHandler(SetStatusOnDelivered);
+
+            aTimerForStopDelivery.Interval = (deltaMilisec + eighteenHoursMilisec) / 600;
```
Comment (Bulczi-CM) [inline CourierService/Program.cs:132]: Packages were supposed to be marked as delivered not 10h after they were send for delivery, but when driver completed the route. Example:

If driver A ships off at 8AM with Shipping List of 5h length, then the packages from his list should be marked as delivered at 1PM, not 6PM.

File: CourierService/Program.cs [hunk 19]
```diff
@@ -103,9 +164,15 @@ void GenerateShipmentList()
            {
                List<Shipment> courierShipmentList = shipmentList.Where(shipment => shipment.DriverId == courier.DriverId).ToList();
 
+               string systemDrivePath = Path.GetPathRoot(Environment.SystemDirectory);
+               string targetPath = $"{systemDrivePath}Shiping_list";
```
Comment (Bulczi-CM) [inline CourierService/Program.cs:168]: For combining paths its good to use Path class, in this case: `var targetPath = Path.Combine(systemDrivePath, "Shiping_list");`

File: CourierService/Program.cs [hunk 20]
```diff
@@ -103,9 +164,15 @@ void GenerateShipmentList()
...
+               Directory.CreateDirectory(targetPath);
                
                var date = _timeService.currentTime().ToString(("MM-dd-yyyy"));
-               var filePath = $@"C:\GitRepository\Courier\Courier\Shipment_List/DriverId{courier.DriverId}_{date}.json";
+
+                string fileName = $"DriverId{courier.DriverId}_{date}.json";
+                string filePath = Path.Combine(targetPath, fileName);
+              //  var filePath = $@"C:\GitRepository\Courier\Courier\Shipment_List/DriverId{courier.DriverId}_{date}.json";
```
Comment (Bulczi-CM) [inline CourierService/Program.cs:175]: remove or uncomment

File: CourierService/Program.cs [hunk 21]
```diff
@@ -203,12 +287,20 @@ void AddCourierCar()
                     Console.WriteLine("\n Min capacity required 500 kg");
                     return;
                 }
-                
+
+                var averageSpeed = _ioHelper.GetUintFromUser("\nAverage speed km/h");
+                if (averageSpeed > 90 || averageSpeed < 30)
+                {
+                    Console.WriteLine("\n Required min averageSpeed 30 km/h, max 90 km/h, try again...");
```
Comment (Bulczi-CM) [inline CourierService/Program.cs:294]: And why is that here? There was no requirements for such a limit :P In my testing set I have WRC and Street Racing Delivery Drivers with average speeds of 120km/h+

**Comment by Bulczi-CM** [comment]
There is no GitIgnore file

**Comment by TomaszNakonieczny80** [comment]
I will send new one

**Comment by TomaszNakonieczny80** [comment]
closed by mistake

File: CourierService/UnityDiContainerProvider.cs [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+namespace Courier
+{
+    public class UnityDiContainerProvider
+    {
+        public IUnityContainer GetContainer()
+        {
+            var container = new UnityContainer();
+
+            container.RegisterType<IMenu, Menu>();
+            container.RegisterType<IMenuItem, MenuItem>();
```
Comment (Bulczi-CM) [inline CourierService/UnityDiContainerProvider.cs:19]: Just wonder how you inject IMenuItem to your Menu

**Review by Bulczi-CM — APPROVED** [review]
```
Tomek - 65%:
1. (25%) Rozwożenie paczek - 15%:
- Predkosc srednia pojazdu nie ma predkosci domyslnes ustawionej tudziez nie ma skryptu ktory zakualizuje juz istniejace wpisy
- Nie bylo wymagania na ograniczenie sredniej przdkosci do wartosci 60-90
- Sposob w jaki wykrywane jest dotarcie paczki do klienta jest niezgodny z wymaganiami. Paczka miala zostac oznaczona jako dostarczona po zakonczeniu rozwozenia przez kuriera, a nie 10h po poczatku jego pracy.
- Mam wrazenie jakby to rozwozenie dzialo sie jakos dziwnie szybko...
2. (10%) Potwierdzenie odebrania paczki przez adresata - 10%:
- Brak uwag
3. (15%) Dependency Injection - 15%:
- Nie wiem po co rejestrujesz MenuItem w kotnenerze
4. (25%) BONUS: Wysyłanie maili - 0%:
- brak
5. (50%) Jakość kodu - 25%:
- Migracji jest tyle ile moglo byc, ale nie dziala update bazy danych
- Jest GitIgnore ale pliki ktore powinny byc ignorowane nadal sa w repo - -5%
- Martwa klasa DateService (nigdie nie wykorzystywana)
- Nazwy niektorych metod sa mylace
- Wielokrotne wywolywanie tej samej metody z tymi samymi parametrami zamiast wyciagniecia raz i przechowania danych w zmiennej
- W jednym miejscu dwa razy wolasz to samo query - raz sprawdzajac cz ydane sa i od razu po nich wyciagajac je. Mozna to zrobic raz  - wyciagasz i sprawdzasz czy cos dostales.
- Obliczenia odleglosci miedzy punktami zasluguja na osobna metode
- Ogromny interface IParcelsService - warto by go podzielic
- Masz jedno cz ydaw miejsca gdzie po pobraniu danych z bazy utrzymujesz niepotrzebnie polaczenie i mielisz dane. Mozna je zamknac i utworzyc celem zapisania, gdy juz skonczysz.
```
